New test framework api structure

AnOctopus opened the pull request:
## Describe your changes
Basic implementation of the new test framework api. It looks big, but a lot of that is because I merely deprecated the original api to give some time to adopt the new one, so several of those large files were copied and modified.


---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

AnOctopus committed:
Hacking something together

AnOctopus committed:
Cleanup

AnOctopus committed:
Move session state handling to TestRunner

AnOctopus committed:
Checkbox test conversion, try WidgetList stuff

AnOctopus committed:
Merge branch 'develop' into feat/test-runner-api-refactor

AnOctopus committed:
Move checkbox test

AnOctopus committed:
Checkbox overloads, demo different accesses

AnOctopus committed:
Use `__call__` to solve all your problems

AnOctopus committed:
Merge pull request #3 from AnOctopus/feat/test-runner-api-refactor-keyed-lookups

Checkbox overloads, demo different accesses

AnOctopus committed:
Add dict-style widget lookup for comparison

AnOctopus committed:
Fix missing overload

Apparently the types on the actual implementation aren't supposed to be used???

AnOctopus committed:
Types, accept slices in getitem, eq

AnOctopus committed:
Remove dead code

AnOctopus committed:
Fragile relative file path magic

AnOctopus committed:
Relative paths work

AnOctopus committed:
Disable script GC in tests

AnOctopus committed:
Query param support

AnOctopus committed:
Restore existing api stuff to `develop` status

AnOctopus committed:
New api stuff lives in `v1`

AnOctopus committed:
Port previous test runner tests

AnOctopus committed:
Access methods for all elements

AnOctopus committed:
Remove rejected api idea implementations

AnOctopus committed:
Fix potential edge case with empty session state

sfc-gh-jcarroll commented on lib/streamlit/testing/v1/local_script_runner.py:
```diff
@@ -0,0 +1,144 @@
+        if runner.script_stopped():
+            return
+
+    # If we get here, the runner hasn't yet completed before our
+    # timeout. Create an error string for debugging.
+    err_string = f"require_widgets_deltas() timed out after {timeout}s)"
```
This is the error string returned the the end user on a timeout, they won't know about require_widgets_deltas - should it be something like `AppTest run timed out after {timeout}s`? Or caught further up the stack and given a friendly error? Also has an extra trailing parenthesis in the string

sfc-gh-jcarroll commented on lib/streamlit/testing/v1/app_test.py:
```diff
@@ -0,0 +1,256 @@
+        """
+        return self._tree.run(timeout)
+
+    @property
+    def main(self) -> Block:
+        return self._tree.main
```
I noticed that equivalent property for `sidebar` is missing on AppTest currently

AnOctopus committed:
Add main and sidebar to AppTest

AnOctopus committed:
Rearrange methods

sfc-gh-jcarroll commented on lib/streamlit/testing/v1/app_test.py:
```diff
@@ -0,0 +1,256 @@
+    @property
+    def main(self) -> Block:
+        return self._tree.main
+
```
Looks like `button` property is also missing on AppTest, I compared the rest and they all seemed to be there

AnOctopus committed:
Add missing `button` property

AnOctopus committed:
User friendly error string

AnOctopus committed:
Documentation update

AnOctopus committed:
Merge branch 'develop' into feat/test-runner-api-refactor

AnOctopus committed:
Inline `_get_widget`

AnOctopus committed:
Fix type issues

AnOctopus committed:
Some test conversions

AnOctopus committed:
Make `.len` a property

AnOctopus committed:
Convert date_input test

AnOctopus committed:
Remove replaced method

AnOctopus committed:
Clean up smoke test

AnOctopus committed:
Move checkbox test back to ett file

AnOctopus committed:
Add some TODOs

AnOctopus committed:
Clarify docstring

lukasmasuch commented on lib/streamlit/testing/v1/app_test.py:
```diff
@@ -0,0 +1,266 @@
+        self._tree = tree
+
+    @classmethod
+    def from_string(cls, script: str, default_timeout: float = 3) -> AppTest:
```
nit: Something to consider could be also to add a `from_function` method since it is quite easy to extract the function source code using `inspect`:

```
import inspect
lines = inspect.getsource(a_function)
```

Especially in our internal tests, using functions would make it easier to work with those tests because the IDE understands it as actual code.

lukasmasuch replied:
For example, it could look something like this:

```python
def test_text_input_interaction():
    def app():
          import streamlit as st
          st.text_area("the label", value=None)
    
    at = AppTest.from_function(app).run()
    text_area = at.text_area[0]
    assert text_area.value is None
 ```

AnOctopus replied:
This is a really nice idea

sfc-gh-jcarroll replied:
+1. feel free to add

sfc-gh-jcarroll replied:
Oh you already did 😅 awesome

lukasmasuch commented on lib/streamlit/testing/v1/util.py:
```diff
@@ -0,0 +1,49 @@
+
+
+@contextmanager
+def patch_config_options(config_overrides: Dict[str, Any]):
```
nit: we can potentially remove the `patch_config_options` and `build_mock_config_get_option` from `tests/testutil.py` and use these versions instead in our tests, or?

AnOctopus replied:
Yep, since that's where I copied from.

lukasmasuch commented on lib/streamlit/testing/v1/app_test.py:
```diff
@@ -0,0 +1,266 @@
+        timed out, if not overridden for an individual `.run()` call.
+        """
+        # TODO: Make this not super fragile
+        stack = traceback.StackSummary.extract(traceback.walk_stack(None))
```
nit: What path is this trying to resolve to? Maybe add a comment here?

AnOctopus replied:
It is trying to find the file the test is in, so we can make file path strings resolve relative to there. But this stack track inspection is so potentially fragile that I'm thinking about giving up on that QoL feature just so it won't randomly break if someone tries to do things slightly differently than I do

lukasmasuch commented on lib/streamlit/testing/v1/app_test.py:
```diff
@@ -0,0 +1,266 @@
+        string in the test itself, without having to create a separate file
+        for it.
+
+        `default_timeout` is the default time in seconds before a script is
```
nit: Since these are user-facing functions, using the numpy docstring format might be better here. Also, if we plan to add this to our docs somehow, this might be required.

lukasmasuch reviewed (approved):
Overall, LGTM 👍 

I'm a bit worried about the aspect that the test features and the actual features might easily diverge if we don't take care of keeping both up-to-date. Not sure if there is a good way to automate checking this; maybe we could add some tests later on to check if at least all commands of the public API are somehow covered. But probably something for later.

sfc-gh-jcarroll commented:
> I'm a bit worried about the aspect that the test features and the actual features might easily diverge if we don't take care of keeping both up-to-date.

@LukasMasuch this is Thiago's biggest concern about the feature too and the ongoing maintenance cost. I have wondered if there's any slick way to get the basic feature working better with less boilerplate code for new commands / elements. Agree we can solve it a little later but any great ideas welcome IMO

lukasmasuch commented on lib/streamlit/testing/v1/element_tree.py:
```diff
@@ -0,0 +1,1506 @@
+    type: str
+    proto: Any = field(repr=False)
+    root: ElementTree = field(repr=False)
+    key: str | None
```
Why do we need to have the `key` here on `Element`? Most read-only elements don't really need a key or?

AnOctopus replied:
See my below answer about `widget_state`

lukasmasuch commented on lib/streamlit/testing/v1/element_tree.py:
```diff
@@ -0,0 +1,1506 @@
+        yield self
+
+    @property
+    def value(self) -> Any:
```
Wouldn't it be possible to move `value` and `widget_state` to `Widget`?

AnOctopus replied:
Non-widget elements also have values. `widget_state` is present on everything so that `ElementTree.get_widget_states()` could handle all node types uniformly. I initially wrote it this way because the nodes are very heterogenous so it felt natural. I normally prefer only defining methods that make actual sense for the specific type, and this might be better written using a separate function that dispatches based on the node's type, rather than relying on subtype polymorphism for dispatch.

lukasmasuch commented on lib/streamlit/testing/v1/app_test.py:
```diff
@@ -0,0 +1,266 @@
+TMP_DIR = tempfile.TemporaryDirectory()
+
+
+class AppTest:
```
Could we extend AppTest somehow from ElementTree so that we don't have to duplicate all command properties?

AnOctopus replied:
We could, though it might make the implementation more confusing

lukasmasuch commented:
@sfc-gh-jcarroll If we want to lower maintenance while still keeping the core value of these interactive unit tests, we could potentially keep (at least the first version) a bit simpler by focusing on modeling the interactive aspects and elements - e.g. setting a widget value - and remove all read-only properties. Specifically, we could expose `proto` as the main way to access read-only properties. Instead of doing `time_input.help`, the user would need to use   `time_input.proto.help` if you want to check something with help. This might not look super nice, but it dramatically lowers this maintenance effort since the protos are autogenerated. We could also rename/expose this as `metadata`, so it looks a bit nicer `time_input.metadata.help`. This will clean up most elements to something like this:
```
@dataclass(repr=False)
class Markdown(Element):
    proto: MarkdownProto

    def __init__(self, proto: MarkdownProto, root: ElementTree):
        self.proto = proto
        self.root = root
        self.type = "markdown"
```

So, we mainly need to keep the commands in sync between the testing framework and the Streamlit API... and not so much the actual properties.

sfc-gh-jcarroll commented:
@LukasMasuch 🤔 As long as `.value` works directly (which seems like the 90% case) I'd be open to something like `.metadata`. Could we do some getattr or subclass magic on the widget object to allow it to work by checking the proto if the attribute exists without listing them explicitly, so the `.metadata` wasn't needed ?

lukasmasuch commented:
@sfc-gh-jcarroll Ah, yes indeed.`.value` could be kept to return the most likely property that represents the value of an element 👍 E.g. `proto.body` for `markdown`.  `getattr` could easily be added in addition, just typing wouldn't be there... but we could always add explicit properties later on if we think are very important for testing. Doing some subclass magic with the proto objects might be tricky from having a quick look at it.

AnOctopus committed:
Add a deprecation warning

AnOctopus commented:
> .metadata proposal

It would work ok for some stuff but a big part of why I went for the wrapper approach rather than expecting direct access of protos is that some proto fields are encoded or otherwise need translation to python types to be friendly to consume, and I've put up with the implementation boilerplate because I didn't come up with an alternative that didn't sacrifice something like good typing. I'm definitely interested in alternatives and will spend some time thinking about how to improve the situation, and I'll add getattr to `Element` as an improvement for stuff that isn't implemented yet

AnOctopus committed:
AppTest.from_function()

An excellent idea by Lukas, though more annoying to implement than I
expected because we don't yet have access to `ast.unparse()`

AnOctopus committed:
Add todo

AnOctopus committed:
Add `__getattr__` fallback to `Element`

AnOctopus committed:
Replace moved functions with re-exports

AnOctopus committed:
Try to make path handling more resilient

AnOctopus committed:
Docstrings!

AnOctopus committed:
Try moving dispatch to isolated function

AnOctopus committed:
Store inner proto for generic elements

AnOctopus committed:
Repr

AnOctopus committed:
Use implementation iheritance to reduce boilerplate

I usually don't reach for it as a tool, so I forgot to consider using it here

AnOctopus committed:
Remove many lines of boilerplate assignments

AnOctopus committed:
Remove redundant field assignments

Yay getattr

AnOctopus committed:
Make `.widget_state()` a property

AnOctopus committed:
Normalize min/max field names

AnOctopus committed:
Remove `.get()` overloads and use type ignores

We prefer to access via properties when the type is known, so the
overloads don't serve much purpose anymore.
Could do it with `cast()`, but it would be a lot more cluttered for no
safety gain.

AnOctopus committed:
Indicate `.widget_state` is for internal use

AnOctopus committed:
Docstrings for `.run()`

AnOctopus committed:
Move root into widget init

AnOctopus committed:
Another `run` docstring

AnOctopus committed:
Update Element doc comment
